source/common/ssl/context_manager_impl.cc

```diff
@@ -1,37 +1,49 @@
+#include "context_impl.h"
 #include "context_manager_impl.h"
 
+#include "common/common/assert.h"
+
 namespace Ssl {
 
-Ssl::ClientContext& ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
-                                                               ContextConfig& config) {
+ContextManagerImpl::~ContextManagerImpl() { ASSERT(contexts_.empty()); }
 
-  Ssl::ClientContext* context = new ClientContextImpl(scope, config);
-  contexts_.emplace_back(context);
-  return *context;
+void ContextManagerImpl::releaseContext(Context* context) {
+  std::unique_lock<std::mutex> lock(contexts_lock_);
+  ASSERT(std::find(contexts_.begin(), contexts_.end(), context) != contexts_.end());
+  contexts_.remove(context);
 }
 
-Ssl::ServerContext& ContextManagerImpl::createSslServerContext(Stats::Scope& scope,
-                                                               ContextConfig& config) {
-  Ssl::ServerContext* context = new ServerContextImpl(scope, config, runtime_);
-  contexts_.emplace_back(context);
-  return *context;
+ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
+                                                            ContextConfig& config) {
+
+  ClientContextPtr context(new ClientContextImpl(*this, scope, config));
+  std::unique_lock<std::mutex> lock(contexts_lock_);
+  contexts_.emplace_back(context.get());
```

Comment:
is this correct?  Won't the ClientContextPtr going out of scope delete the new ClientContextImpl?